synapse/handlers/device.py

```diff
@@ -678,6 +681,18 @@ async def store_dehydrated_device(
         )
         if old_device_id is not None:
             await self.delete_devices(user_id, [old_device_id])
+
+        # we do this here to avoid a circular import
+        if self.hs.config.worker.worker_app is None:
+            # if main process
+            key_uploader = self.hs.get_e2e_keys_handler().upload_keys_for_user
+        else:
+            # if worker process
+            key_uploader = ReplicationUploadKeysForUserRestServlet.make_client(self.hs)
+
+        # if keys are provided store them
+        if device_keys:
```

Comment:
Correct me if I'm wrong, and my Python-foo might not be the best anymore, but I don't think this will ensure that we can't have a dehydrated device without device keys.

For this guaranteed to be upheld, the device and its keys need to be written to the database in a single database transaction. We still have multiple calls and multiple write operations that can independently fail, albeit, they might now be in a nicer place.

Reply:
Your python-foo is fine, I just misunderstood that this should be a single database transaction, and I see what you mean about just making the problem more convenient.

Reply:
Storing the keys in the same DB transaction as storing the dehydrated device was a little trickier but I gave it a shot and have updated this PR to do so.